### PR TITLE
1.4.0 updates to 3PAR/Primera and lab guide updates

### PR DESCRIPTION
**Pull request by c-snell**

### CODE REVIEW

File: docs/container_storage_provider/hpe_3par_primera/index.md [hunk 1]
```diff
@@ -11,16 +11,16 @@ The HPE 3PAR and Primera Container Storage Provider integrates as part of the [H
 
 Always check the corresponding CSI driver version in [compatibility and support](../../csi_driver/index.md#compatibility_and_support) and [SPOCK](#spock) for latest support matrix for the HPE 3PAR and Primera Container Storage Provider.
 
-|  CSI   | Linux OS | OpenShift | Kubernetes | 3PAR and Primera OS |
-| ------ | -------- | --------- | ---------- | ------------------- |
-| v1.3.0 | - CentOS: 7.7 <br /> - RHEL: 7.6, 7.7 / RHCOS | OpenShift 4.2/4.3 with RHEL 7.6 or 7.7 or RHCOS as worker nodes| K8s 1.16-1.18 | - 3PAR OS: 3.3.1 (FC & iSCSI) <br /> - Primera OS: 4.0.0, 4.1.0 (FC only), 4.2.0 |
+|  CSI   | Transport Protocols | &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; Linux OS &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;  | Red Hat OpenShift/SUSE CaaSP | &nbsp; &nbsp; &nbsp; Kubernetes &nbsp; &nbsp; &nbsp; | 3PAR and Primera OS |
+| ------ | ------------------- |-------- | --------- | ---------- | ------------------- |
+| v1.4.0 | iSCSI & FC | - CentOS: 7.6+, 8.1 <br /> - RHEL: 7.6+, 8.1 <br /> - RHCOS | - OpenShift 4.4, 4.6 <br /> - SUSE CaaSP 4.2 | - K8s 1.17-1.20 <br /> - Google Anthos GKE 1.4 | - 3PAR OS: 3.3.1+ <br /> - Primera OS: 4.0+ |
```
Comment (datamattsson) [inline docs/container_storage_provider/hpe_3par_primera/index.md:16]: This table is a bit of an eye sore. Could this be formatted to with the page estate better. Also:

- SUSE CaaSP 4.2 is listed as a solution but we're missing SLES in the host OS column
- Anthos is listed under "Kubernetes" when it's a partner solution just like OCP and CaaSP
- The leading dashes on the entries just steal horizontal estate

Maybe just merge "Kubernetes" and "Red Hat OpenShift/SUSE CaaSP" to "Platforms" and enumerate them.

File: docs/container_storage_provider/hpe_3par_primera/index.md [hunk 2]
```diff
@@ -229,31 +236,53 @@ These parameters are applicable only for replication. Both parameters are mandat
...
+### QoS StorageClass parameter
 
-## Support
+In the HPE Primera or 3PAR Storage system, the QoS rules are applied to a volume set. To use an existing volume set with QoS rules on a `PersistentVolumeClaim`, set the `qosName` parameter within a `StorageClass` to the name of the existing HPE Primera or 3PAR volume set.
+
+| Parameter          | Option  | Description |
+| ------------------ | ------- | ----------- |
+| qosName      | Text         | Name of the HPE Primera or 3PAR volume set which has QoS rules. This parameter is optional. If specified, the `PersistentVolumeClaim` will be associated with the HPE Primera or 3PAR volume set, for purposes of applying the QoS rules. |
+
+### VolumeGroupClass QoS parameters
```
Comment (datamattsson) [inline docs/container_storage_provider/hpe_3par_primera/index.md:273]: No SnapshotGroupClass parameters?

File: docs/container_storage_provider/hpe_3par_primera/index.md [hunk 3]
```diff
@@ -229,31 +236,53 @@ These parameters are applicable only for replication. Both parameters are mandat
...
+In the HPE CSI Driver 1.4.0, a volume set with QoS settings can be created dynamically using the QoS parameters for the `VolumeGroupClass`. The following parameters are available for a `VolumeGroup` on the HPE Primera or 3PAR Storage system. Learn more about `VolumeGroups` in the [provisioning concepts documentation](../../csi_driver/using.md#volume_groups).
+
+| Parameter          | String  | Description |
+| ------------------ | ------- | ----------- |
+| priority        | Text    | QoS priority. Example values: "low", "normal", "high"|
+| ioMinGoal        | Text    | I/O-per-second minimum goal. Example values: "300" |
+| ioMaxLimit        | Text    | I/O-per-second maximum limit. Example values: "10000" |
+| bwMinGoalKb        | Text    | Bandwidth minimum goal in kilobytes per second. Example values: "300" |
+| latencyGoal        | Text    | Bandwidth maximum limit in kilobytes per second. Example values: "30000" |
+| domain        | Text    | The 3PAR/Primera domain, with which the volume group and related objects are associated with. Example values: "testdomain" |
```
Comment (datamattsson) [inline docs/container_storage_provider/hpe_3par_primera/index.md:284]: Not my area of expertise but isn't the customer facing name "virtual domain"? These seemed to be copied verbatim from engineering.

File: docs/container_storage_provider/hpe_3par_primera/index.md [hunk 4]
```diff
@@ -229,31 +236,53 @@ These parameters are applicable only for replication. Both parameters are mandat
...
+
+### VolumeGroupClass QoS parameters
+
+In the HPE CSI Driver 1.4.0, a volume set with QoS settings can be created dynamically using the QoS parameters for the `VolumeGroupClass`. The following parameters are available for a `VolumeGroup` on the HPE Primera or 3PAR Storage system. Learn more about `VolumeGroups` in the [provisioning concepts documentation](../../csi_driver/using.md#volume_groups).
+
+| Parameter          | String  | Description |
+| ------------------ | ------- | ----------- |
+| priority        | Text    | QoS priority. Example values: "low", "normal", "high"|
+| ioMinGoal        | Text    | I/O-per-second minimum goal. Example values: "300" |
+| ioMaxLimit        | Text    | I/O-per-second maximum limit. Example values: "10000" |
```
Comment (datamattsson) [inline docs/container_storage_provider/hpe_3par_primera/index.md:281]: I/O-per-second is usually abbreviated as IOPS.

File: docs/container_storage_provider/hpe_3par_primera/index.md [hunk 5]
```diff
@@ -229,31 +236,53 @@ These parameters are applicable only for replication. Both parameters are mandat
...
+| bwMinGoalKb        | Text    | Bandwidth minimum goal in kilobytes per second. Example values: "300" |
+| latencyGoal        | Text    | Bandwidth maximum limit in kilobytes per second. Example values: "30000" |
```
Comment (datamattsson) [inline docs/container_storage_provider/hpe_3par_primera/index.md:283]: parameter name does not match text.

File: docs/learn/persistent_storage/index.md [hunk 6]
```diff
@@ -445,29 +304,32 @@ Or modify the webpage:
...
+To get started with the deployment, the HPE CSI Driver is deployed using industry standard means, either a Helm chart or an Operator. For this tutorial, we will be using Helm to the deploy the HPE CSI driver for Kubernetes.
 
 The official Helm chart for the HPE CSI Driver for Kubernetes is hosted on [Artifact Hub](https://artifacthub.io/packages/helm/hpe-storage/hpe-csi-driver). There, you will find the configuration and installation instructions for the chart.
 
 
-### Installing the chart
+### Installing the Helm chart
 
-To install the chart with the name hpe-csi, add the HPE CSI Driver for Kubernetes helm repo.
+To install the chart with the name `hpe-csi`, add the HPE CSI Driver for Kubernetes helm repo.
```
Comment (datamattsson) [inline docs/learn/persistent_storage/index.md:322]: Helm is a name.

File: docs/learn/persistent_storage/index.md [hunk 7]
```diff
@@ -481,36 +343,40 @@ kubectl get pods --all-namespaces -l 'app in (nimble-csp, primera3par-csp, hpe-c
...
-kube-system   primera3par-csp-66f775b555-sfmnp      1/1       Running   0          5m
+NAME                                  READY   STATUS    RESTARTS   AGE
+hpe-csi-controller-66fc6544f8-mnllq   9/9     Running   0          77s
+hpe-csi-node-gqjnq                    2/2     Running   0          77s
+nimble-csp-5d4c9fc5b6-sxqmm           1/1     Running   0          77s
+primera3par-csp-94b9c4978-4xv86       1/1     Running   0          77s
 ```
 
-If all of the components show in Running state, then the HPE CSI driver for Kubernetes and the corresponding Container Storage Providers have been successfully deployed.
+If all of the components show in Running state, then the HPE CSI driver for Kubernetes and the corresponding Container Storage Providers (CSP) for Nimble, Primera or 3PAR have been successfully deployed.
```
Comment (datamattsson) [inline docs/learn/persistent_storage/index.md:353]: **D**river

**Review by datamattsson — CHANGES_REQUESTED** [review]
A few nits. Did you investigate if we can pull example images (nginx) from quay.io instead of docker.io?

File: docs/container_storage_provider/hpe_3par_primera/index.md (hunk 5 above)
Comment (c-snell) [reply docs/container_storage_provider/hpe_3par_primera/index.md:283]: The parameter and verbage is correct as a volume set must be created and the QoS applied to the vvset. The parameter is then set to the vvset and volumes put into the vvset get the QoS settings applied to them. The parameter name doesn't reflect what is to be named here which has been fixed with Volume Group QoS.

File: docs/learn/persistent_storage/index.md [hunk 8]
```diff
@@ -690,3 +690,40 @@ Access the admin console at: **http://127.0.0.1/admin** using the **"admin/admin
...
+helm uninstall my-hpe-csi-driver -n hpe-storage
+```
+
+It takes a couple minutes to cleanup the objects from the CSI driver. You can check the status:
+
+```
+watch kubectl get all -n hpe-storage
+```
+
+Once everything is removed, **Ctrl+C** to exit and finally you can remove the `namespace`.
```
Comment (datamattsson) [inline docs/learn/persistent_storage/index.md:725]: When referencing a Kubernetes `kind`, we should use the correct capitalization of the `kind`. Which is `Namespace` in this case.

**Review by datamattsson — CHANGES_REQUESTED** [review]
Mooar nits!

**Review by datamattsson — APPROVED** [review]
LGTM

**Pull request merged**